lc0/src/neural/network_cudnn.cu

```diff
@@ -862,44 +887,8 @@ class CudnnNetwork : public Network {
   float *tensor_mem_[3];
   float *scratch_mem_;
 
-  // GPU accessible system memory used to store inputs and outputs
-  // avoids copies system <-> device memory
-  std::map<std::thread::id, uint64_t *> input_masks_mem_;
-  std::map<std::thread::id, float *> input_val_mem_;
-  std::map<std::thread::id, float *> op_policy_mem_;
-  std::map<std::thread::id, float *> op_value_mem_;
-
-  // GPU pointers for the above allocations
-  std::map<std::thread::id, uint64_t *> input_masks_mem_gpu_;
-  std::map<std::thread::id, float *> input_val_mem_gpu_;
-  std::map<std::thread::id, float *> op_policy_mem_gpu_;
-  std::map<std::thread::id, float *> op_value_mem_gpu_;
-
-  void allocInputsOutputsMem(std::thread::id tid) {
-    reportCUDAErrors(cudaHostAlloc(
-        &input_masks_mem_[tid], kMaxBatchSize * kInputPlanes * sizeof(uint64_t),
-        cudaHostAllocMapped));
-    reportCUDAErrors(cudaHostGetDevicePointer(&input_masks_mem_gpu_[tid],
-                                              input_masks_mem_[tid], 0));
-
-    reportCUDAErrors(cudaHostAlloc(&input_val_mem_[tid],
-                                   kMaxBatchSize * kInputPlanes * sizeof(float),
-                                   cudaHostAllocMapped));
-    reportCUDAErrors(cudaHostGetDevicePointer(&input_val_mem_gpu_[tid],
-                                              input_val_mem_[tid], 0));
-
-    reportCUDAErrors(cudaHostAlloc(
-        &op_policy_mem_[tid], kMaxBatchSize * kNumOutputPolicy * sizeof(float),
-        cudaHostAllocMapped));
-    reportCUDAErrors(cudaHostGetDevicePointer(&op_policy_mem_gpu_[tid],
-                                              op_policy_mem_[tid], 0));
-
-    reportCUDAErrors(cudaHostAlloc(&op_value_mem_[tid],
-                                   kMaxBatchSize * sizeof(float),
-                                   cudaHostAllocMapped));
-    reportCUDAErrors(cudaHostGetDevicePointer(&op_value_mem_gpu_[tid],
-                                              op_value_mem_[tid], 0));
-  }
+  mutable std::mutex inputs_outputs_lock_;
+  std::list<InputsOutputs *> free_inputs_outputs_;
```

Comment:
You can make it `std::list<std::unique_ptr<InputsOutputs>>` (and also `std::unique_ptr<InputsOutputs> CudnnNetworkComputation::inputs_outputs_`)
then create with `std::make_unique<InputsOutputs>()`  (instead of `new`)
and move around with `std::move(val)`.
e.g.:
```cpp
// std::unique_ptr<InputsOutputs> io
free_inputs_outputs_.push_back(std::move(io));
```
That we there's no need to remember to release it in destructor, and ownership is always explicit.